frontend/src/components/pages/TranslationPage.tsx

```diff
@@ -0,0 +1,220 @@
+  const addIcon = (): JSX.Element | null => {
+    const randomNum = Math.floor(Math.random() * 3);
+
+    switch (randomNum) {
+      case 0:
+        return <img src={CheckmarkIcon} alt="Approved line" />;
+      case 1:
+        return <img src={CommentIcon} alt="Line with feedback" />;
+      default:
+        return null;
+    }
+  };
+
+  const [updateTranslation] = useMutation<{
+    updateStoryTranslationContentById: UpdateTranslationResponse;
+  }>(UPDATE_TRANSLATION);
+  const onChangeTranslationContent = async (
+    storyTranslationContentId: number,
+    newContent: string,
+  ) => {
+    try {
+      const result = await updateTranslation({
+        variables: {
+          storyTranslationContentId,
+          newTranslationContent: newContent,
+        },
+      });
+
+      const lineIndex =
+        result.data?.updateStoryTranslationContentById.story.lineIndex;
+      if (lineIndex !== undefined) {
+        // this is update only, so id already set
+        // not convinced this is the best strateg, this will re-redner the page
+        // and the box that has already been updated right?
+        // should only update whatever is storingi nfo without re-rendering if possible
+        const updatedContentArray = [...translatedStoryLines];
+        const index = arrayIndex(lineIndex);
+        updatedContentArray[index].translatedContent = newContent;
+
+        setTranslatedStoryLines(updatedContentArray);
+      } else {
+        handleError("Unable to save translation.");
+      }
+    } catch (err) {
+      handleError(err ?? "Error occurred, please try again.");
+    }
+  };
+
+  useQuery(GET_STORY_CONTENTS(storyId, storyTranslationId), {
+    fetchPolicy: "cache-and-network",
+    onCompleted: (data) => {
+      const storyContent = data.storyById.contents;
+      const translatedContent = data.storyTranslationById.translationContents;
+
+      const contentArray: StoryLine[] = [];
+      // would callign directly onto state object rather than this array re-render
+      // expect an ordered response
+      storyContent.forEach(({ id, content, lineIndex }: Content) => {
+        contentArray.push({
+          lineIndex,
+          originalContent: content,
+          storyContentId: id,
+        });
+      });
+      translatedContent.forEach(({ id, content, lineIndex }: Content) => {
+        // why is "" registering as undefined
+        contentArray[lineIndex].translatedContent = content ?? "";
+        contentArray[lineIndex].storyTranslationContentId = id;
+      });
+
+      setTranslatedStoryLines(contentArray);
+    },
+  });
+
+  // TODO not in order? maybe fix it via css -> is it possible since not guaranteed order (what div properties needed)
+  // array might help then
+  // pagination will give us troubles in this format -> will need to subtract firsti ndex or smth
+  // is that super inefficient
+
+  // double loop from response?
+  const storyCells = translatedStoryLines.map((storyLine: StoryLine) => {
```

Comment:
My big (?) here is whether I can avoid looping so much -> right now, I'm looping
- once over the storyContents response
- once over the storyTranslationContents response
- once over the combined object state array (which has everything from the above)

That's not so painful for my 10-element array, but could get gross for larger stories?

Reply:
Not sure how zipping would look. Without zipping, I'm thinking the options would be
- 2 sorts (both response arrays) + one iteration (go through both together by index)
- 1 sort (first response array) + 2 iterations (go through each separately)

I think second should be more efficient?

Reply:
I agree, second is more efficient